Firestore/core/src/firebase/firestore/local/local_serializer.cc

```diff
@@ -35,97 +35,79 @@ namespace firestore {
 namespace local {
 
 using core::Query;
+using model::Document;
+using model::MaybeDocument;
+using model::NoDocument;
 using model::ObjectValue;
 using model::SnapshotVersion;
 using nanopb::Reader;
 using nanopb::Tag;
 using nanopb::Writer;
 using util::Status;
 
-Status LocalSerializer::EncodeMaybeDocument(
-    const model::MaybeDocument& document,
-    std::vector<uint8_t>* out_bytes) const {
-  Writer writer = Writer::Wrap(out_bytes);
-  EncodeMaybeDocument(&writer, document);
-  return writer.status();
-}
-
 void LocalSerializer::EncodeMaybeDocument(
-    Writer* writer, const model::MaybeDocument& maybe_doc) const {
+    Writer* writer, const MaybeDocument& maybe_doc) const {
   switch (maybe_doc.type()) {
-    case model::MaybeDocument::Type::Document:
+    case MaybeDocument::Type::Document:
       writer->WriteTag(
           {PB_WT_STRING, firestore_client_MaybeDocument_document_tag});
       writer->WriteNestedMessage([&](Writer* writer) {
-        EncodeDocument(writer, static_cast<const model::Document&>(maybe_doc));
+        EncodeDocument(writer, static_cast<const Document&>(maybe_doc));
       });
       return;
 
-    case model::MaybeDocument::Type::NoDocument:
+    case MaybeDocument::Type::NoDocument:
       writer->WriteTag(
           {PB_WT_STRING, firestore_client_MaybeDocument_no_document_tag});
       writer->WriteNestedMessage([&](Writer* writer) {
-        EncodeNoDocument(writer,
-                         static_cast<const model::NoDocument&>(maybe_doc));
+        EncodeNoDocument(writer, static_cast<const NoDocument&>(maybe_doc));
       });
       return;
 
-    case model::MaybeDocument::Type::Unknown:
+    case MaybeDocument::Type::Unknown:
       // TODO(rsgowman)
       abort();
   }
 
   UNREACHABLE();
 }
 
-std::unique_ptr<model::MaybeDocument> LocalSerializer::DecodeMaybeDocument(
+std::unique_ptr<MaybeDocument> LocalSerializer::DecodeMaybeDocument(
     Reader* reader) const {
-  if (!reader->status().ok()) return nullptr;
-
-  std::unique_ptr<model::MaybeDocument> result;
+  std::unique_ptr<MaybeDocument> result;
 
-  while (reader->bytes_left()) {
-    Tag tag = reader->ReadTag();
-    if (!reader->status().ok()) return nullptr;
-
-    // Ensure the tag matches the wire type
-    switch (tag.field_number) {
+  while (reader->good()) {
+    switch (reader->ReadTag().field_number) {
```

Comment:
I'm for option 3. Tag vs field number is a distinction without any practical difference. That we happen to read the wire type as a part of reading the "tag" and then use it implicitly in the next operation is just an implementation detail.

Basically I'm pushing you toward making this opinionated and minimal for the user of the API rather than fully flexible and technically correct.